lib/ansible/module_utils/facts/collector.py

```diff
@@ -248,19 +336,18 @@ def collector_classes_from_gather_subset(all_collector_classes=None,
                                           aliases_map=aliases_map,
                                           platform_info=platform_info)
 
-    # TODO: can be a set()
-    seen_collector_classes = []
+    unresolved = find_unresolved_requires(collector_names, all_fact_subsets)
 
-    selected_collector_classes = []
+    new_names = resolve_requires(unresolved, all_fact_subsets)
+    collector_names.update(new_names)
 
-    for collector_name in collector_names:
-        collector_classes = all_fact_subsets.get(collector_name, [])
+    dep_map = build_dep_data(collector_names, all_fact_subsets)
 
-        # TODO? log/warn if we dont find an implementation of a fact_id?
+    ordered_deps = tsort(dep_map)
+    collector_names = [x[0] for x in ordered_deps]
```

Comment:
MAYBE: add a convenience method for this. Probably will if I move tsort and friends to a separate file.